### ISSUE TITLE
Better handling of missing fonts

### ISSUE THREAD

**Issue by trych** [basiljs/basil.js#119]
If a font is called by `b.textFont()` that does not exist on the current system the entire script stops. However, when teaching/learning basil.js in class this is a situation that happens *all. the. time.*

People write little scripts that use a specific font and if someone else wants to try the scripts then either they have to get the font file from somewhere or they have to comment out all the font lines. This situation happens so much that it becomes really annoying to always comment out the lines (and then to remember to reset it in the end).

I think it would be nicer if the script would still execute, fall back to the default font and instead give a strong warning (maybe even with an alert window?) that font xyz is missing and that the visual result of the script will likely be different than intended.

**Comment by fabianmoronzirfas** [comment]
I totally agree. The execution should not be stopped.

**Comment by b-g** [comment]
Same here, it should not stop but be verbose about that a font is missing -> warning in console

**Comment by trych** [comment]
Ok, I just was thinking about making this more prominent than just a warning in the console, but I have to admit that stopping the execution with an alert is not very elegant. 😉

How about a warning at the very end of the console message:

`### Basil Warning -> textFont(), the fonts "font1", "font2" and "font3" are missing. The result of the script will likely look different than intended.`

That way at least it gets not buried between other console logs and it is nice to have all fonts collected.

**Comment by fabianmoronzirfas** [comment]
This depends on how you like to log things and is opinionated. I like to have my log messages in the order in which they occur. This makes tracing the error easier. What about @b-g or @ffd8 ? (Man I also need a shorter name. So much to type on iOS:) 

Maybe also ask the attendees of your seminar and explain both pro and cons beforehand.

**Comment by trych** [comment]
> I like to have my log messages in the order in which they occur.

That's a good point.

How about this then: Report missing fonts in order of appearance (or lack thereof 😉) and then add a summary at the end.

So it would be:

```js
### Basil Warning -> textFont(), "font1" is missing.

### Basil Warning -> textFont(), "font2" is missing.

### Basil Warning -> textFont(), "font3" is missing.

### Basil Warning -> textFont(), the fonts "font1", "font2" and "font3" are missing. The result of the script will likely look different than intended.
```

Alternative would be

```js
### Basil Warning -> textFont(), "font1" is missing. The result of the script will likely look different than intended.

### Basil Warning -> textFont(), "font2" is missing. The result of the script will likely look different than intended.

### Basil Warning -> textFont(), "font3" is missing. The result of the script will likely look different than intended.
```

I prefer the first version.

**Comment by b-g** [comment]
Good point! I would rather go for the most simple one = 2 ...

**Comment by fabianmoronzirfas** [comment]
+1 on V2

**Comment by b-g** [comment]
And  just `Basil Warning -> textFont(), "font3" is missing.` would be enough for me

**Comment by trych** [comment]
Hm, I am just thinking of cases where people download a script, execute it and then the result has overset text and looks weird. If they have not much experience with basil or scripting it might be nice to give a pointer that the result looks differently than expected.

If you don't check the console you might never even find out that something is wrong.

Should we still risk this?

**Comment by ffd8** [comment]
Sounds great. Also agree using console.log when/where it happens is ideal.
Perhaps happy medium for detail and length could be:

`Basil Warning -> textFont(), "font3" missing. Using default font.`

[Default font](https://forums.adobe.com/thread/849505), would be whatever font is currently active when no TF is active:
`app.activeDocument.textDefaults.appliedFont`

**Comment by fabianmoronzirfas** [comment]
>If you don't check the console you might never even find out that something is wrong.  
>Should we still risk this?  

I think it is okay. It is still a programming learning platform and not for bullet proofed endusers distribution. If somebody uses it for distribution he should be able to check whether his fonts are there or not.

**Comment by trych** [comment]
> If somebody uses it for distribution he should be able to check whether his fonts are there or not.

I'm rather thinking of online sample scripts, like on the [page of your past course](https://fabianmoronzirfas.github.io/Typography-And-Automation/) for example.

But having said that, I think ffd8's suggestion is an ideal compromise. It gets the point across, but is still short enough not to clutter the console.

**Comment by fabianmoronzirfas** [comment]
> like on the page of your past course for example.
This was in 2012. Damn you never forgetting Internet 😆 
I know the problem but still think it should not be over engineered

**Comment by trych** [comment]
But what do you think about ff8d's suggestion? Would that work for you?

**Comment by fabianmoronzirfas** [comment]
Totally. There should be a fallback. Wait I'll give it a thumbs up 😉

**Comment by trych** [comment]
Ok, I think then we just go with this one. Yes, fallback should be the default font which we need to retrieve at the beginning of the script anyways as proposed in #85.

**Comment by trych** [comment]
This was resolved in #201.
